torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +120,98 @@ def _warn_on_overridden_mixed_precision(
+        for child_module_name, child_module in module.named_children():
+            if child_module is None:  # only for overrides of `named_children()`
+                continue
+            if child_module not in visited_modules:
+                visited_modules.add(child_module)
+                if module_name != "":
+                    child_module_name = module_name + "." + child_module_name
+                stack.append((child_module_name, child_module))
+    reverse_topo_sorted_modules = reversed(topo_sorted_named_modules)
+    visited_modules.clear()
+    for module_name, module in reverse_topo_sorted_modules:
+        if module in modules_to_wrap:
+            param_to_fqn = _get_param_to_fqn(
+                module, ignored_params, visited_modules, module_name
+            )
+            frozen_param_fqns: List[str] = []
+            nonfrozen_param_fqns: List[str] = []
+            for param, fqn in param_to_fqn.items():
+                if param.requires_grad:
+                    nonfrozen_param_fqns.append(fqn)
+                else:
+                    frozen_param_fqns.append(fqn)
+            if len(frozen_param_fqns) > 0 and len(nonfrozen_param_fqns) > 0:
+                msg = f"{module_name} has both parameters with requires_grad=True and False."
+                if strict:
+                    msg += " FSDP does not support wrapping such modules.\n"
+                else:
+                    msg += (
+                        " FSDP does not recommend wrapping such modules since "
+                        "the gradient memory usage will be higher than expected.\n"
+                    )
+                msg += (
+                    f"The following parameters have requires_grad=True:\n{nonfrozen_param_fqns}\n"
+                    f"The following parameters have requires_grad=False:\n{frozen_param_fqns}"
+                )
+                if strict:
+                    raise ValueError(msg)
+                else:
+                    warnings.warn(msg)
+
+
+def _get_param_to_fqn(
+    root_module: nn.Module,
+    ignored_params: Set[nn.Parameter],
+    visited_modules: Set[nn.Module],
+    root_prefix: str,
+) -> Dict[nn.Parameter, str]:
+    """
+    NOTE: This function differs from the ``_get_param_to_fqn()`` used for
+    ``rekey_optim_state_dict()``. Here, we rely on the keys in the dict
+    being exactly the parameters that would be managed by ``root_module`` given
+    ``visited_modules``, which depends on the target modules to wrap.
+    """
+    param_to_fqn: Dict[nn.Parameter, str] = {}
+    # Run BFS (or any tree traversal works)
+    queue = collections.deque([(root_module, root_prefix)])
+    visited_modules.add(root_module)
+    while queue:
+        module, prefix = queue.popleft()
+        for param_name, param in module.named_parameters(recurse=False):
+            if param not in ignored_params:
+                fqn = param_name if prefix == "str" else prefix + "." + param_name
```

Comment:
Why there is a magic str `"str"`?

Reply:
Thanks for the catch! I am not sure where that came from -- it should just be `""`.